### PR TITLE
Cleanup: unify set implementations behind a singular Interface type

### PR DESCRIPTION
**Pull request by hmoog**
This PR introduces a generalized Set type as an interface and start to flatten the ds package to aggregate all major types in a singular package.

### CODE REVIEW

File: ds/set_impl.go [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+	defer s.applyMutex.Unlock()
+
+	removedElements = NewSet(s.ToSlice()...)
+	s.Clear()
+
+	elements.Range(func(element ElementType) {
+		s.Set(element, types.Void)
+	})
+
+	return removedElements
```
Comment (karimodm) [inline ds/set_impl.go]: Shouldn't this return the removedElements - removedElements.Intersect(elements)?

File: ds/set_impl.go [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+	return s.readableSet
+}
+
+// endregion ///////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// region readableSet //////////////////////////////////////////////////////////////////////////////////////////////////
+
+// readableSet is the standard implementation of the ReadableSet interface.
+type readableSet[T comparable] struct {
+	*orderedmap.OrderedMap[T, types.Empty] `serix:"0"`
```
Comment (karimodm) [inline ds/set_impl.go:132]: Would it be better to use a shrinkingmap here for the writeable counterpart as elements can be deleted?

**Review by karimodm — CHANGES_REQUESTED** [review]
Can you add test coverage for Apply (and setMutations) and Replace?

File: ds/set_impl.go (hunk 1 above)
Comment (hmoog) [reply ds/set_impl.go]: It should actually return the "previousValues" and no it should not intersect them as we need the full set to construct the mutations.

I renamed the return value to indicate that it returns all of the previousElements.

File: ds/set_impl.go (hunk 2 above)
Comment (hmoog) [reply ds/set_impl.go:132]: The OrderedMap brings Encode and Decode methods and uses a shrinking map internally, so for convenience and to not have to implement the methods manually I would keep it like it is.

**Review by karimodm — APPROVED** [review]

**Pull request merged**